cmd/geth/snapshot.go

```
@@ -0,0 +1,409 @@
+	if err != nil {
+		utils.Fatalf("Failed to open snapshot tree %v", err)
+	}
+	if ctx.NArg() > 1 {
+		utils.Fatalf("Too many arguments given")
+	}
+	var targetRoot common.Hash
+	if ctx.NArg() == 1 {
+		targetRoot, err = parseRoot(ctx.Args()[0])
+		if err != nil {
+			utils.Fatalf("Failed to resolve state root %v", err)
+		}
+	}
+	if err = pruner.Prune(targetRoot); err != nil {
+		utils.Fatalf("Failed to prune state %v", err)
+	}
+	return nil
+}
+
+func verifyState(ctx *cli.Context) error {
+	stack, _ := makeConfigNode(ctx)
+	defer stack.Close()
+
+	chain, chaindb := utils.MakeChain(ctx, stack, true)
+	defer chaindb.Close()
+
+	snaptree, err := snapshot.New(chaindb, trie.NewDatabase(chaindb), 256, chain.CurrentBlock().Root(), false, false, false)
+	if err != nil {
+		log.Crit("Failed to open the snapshot tree", "error", err)
+	}
+	if ctx.NArg() > 1 {
+		log.Crit("Too many arguments given")
+	}
+	var root = chain.CurrentBlock().Root()
+	if ctx.NArg() == 1 {
+		root, err = parseRoot(ctx.Args()[0])
+		if err != nil {
+			utils.Fatalf("Failed to resolve state root %v", err)
+		}
+	}
+	if err := snaptree.Verify(root); err != nil {
+		log.Crit("Failed to verfiy state", "error", err)
+	} else {
+		log.Info("Verified the state")
+	}
+	return nil
+}
+
+// traverseState is a helper function used for pruning verification.
+// Basically it just iterates the trie, ensure all nodes and associated
+// contract codes are present.
+func traverseState(ctx *cli.Context) error {
+	stack, _ := makeConfigNode(ctx)
+	defer stack.Close()
+
+	chain, chaindb := utils.MakeChain(ctx, stack, true)
+	defer chaindb.Close()
+
+	if ctx.NArg() > 1 {
+		log.Crit("Too many arguments given")
+	}
+	// Use the HEAD root as the default
+	head := chain.CurrentBlock()
+	if head == nil {
+		log.Crit("Head block is missing")
```

Better to return the error here, so the chain and dbs ares properly closed